### PR TITLE
STScI DSI-68: Expose direct HST S3 urls

### PR DESCRIPTION
**Pull request by cam72cam**
This will make integration with tools like Lambda simpler.

```
from astroquery.mast import Observations
obsTable = Observations.query_object("M101",radius=0.02)
hstObs = obsTable[obsTable["obs_collection"]=="HST"]
products = Observations.get_product_list(hstObs[:4])
Observations.enable_s3_hst_dataset()

Observations.get_hst_s3_uris(products)
['s3://stpubdata/hst/public/ib3p/ib3p11010/ib3p11010_asn.fits', 's3://stpubdata/hst/public/ib3p/ib3p11010/ib3p11010_drc.fits', 's3://stpubdata/hst/public/ib3p/ib3p11010/ib3p11010_drz.fits', 's3://stpubdata/hst/public/icae/icae03010/icae03010_asn.fits', 's3://stpubdata/hst/public/icae/icae03010/icae03010_drc.fits', 's3://stpubdata/hst/public/icae/icae03010/icae03010_drz.fits', 's3://stpubdata/hst/public/icae/icae03020/icae03020_asn.fits', 's3://stpubdata/hst/public/icae/icae03020/icae03020_drc.fits', 's3://stpubdata/hst/public/icae/icae03020/icae03020_drz.fits', 's3://stpubdata/hst/public/icae/icae03030/icae03030_asn.fits', 's3://stpubdata/hst/public/icae/icae03030/icae03030_drc.fits', 's3://stpubdata/hst/public/icae/icae03030/icae03030_drz.fits']

Observations.get_hst_s3_uri(products[0])
's3://stpubdata/hst/public/ib3p/ib3p11010/ib3p11010_asn.fits'

Observations.get_hst_s3_uri(products[0], includeBucket=False)
'hst/public/ib3p/ib3p11010/ib3p11010_asn.fits'
```

CC @arfon

### CODE REVIEW

File: astroquery/mast/core.py [hunk 1]
```diff
@@ -1354,33 +1359,43 @@ def _download_from_s3(self, dataProduct, localPath, cache=True):
...
-                        raise
+        # Unfortunately our file placement logic is anything but sane
+        # We put files in folders that don't make sense
+        for ch in (string.digits + string.ascii_lowercase):
+            # The last char of the obs_folder (observation id) can be any lowercase or numeric char
+            insane_obs = obs_id[:-1] + ch
+            insane_path = os.path.join("hst", "public", insane_obs[:4], insane_obs, filename)
+            paths += [insane_path]
+
+        for path in paths:
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply astroquery/mast/core.py]: It is fairly expensive, though I think at least 75% of the time the sane path works as expected.

The AWS API is pretty good with high request volumes.  When I wrote some particularly bad code for another project we started getting rate limit warnings and errors, but boto passes those up as either log messages or raises exceptions depending on the severity.  Then again, I was really abusing the API from hundreds of machines in parallel...

As far as I am aware there is no better way to handle this.

[the remark this answers is not in the document]
Comment (keflavich) [reply astroquery/mast/core.py]: OK, sounds good.

**Comment by arfon** [comment]
@cam72cam - this looks great. I wonder if we should add the option to have the full URL e.g. `http(s)://s3.amazonaws.com/<bucket>/<object>`

**Comment by bsipocz** [comment]
@cam72cam - I think it's probably fine without, but feel free to add a changelog entry if you think it's useful.

**Comment by cam72cam** [comment]
@bsipocz I added a changelog entry

**Comment by bsipocz** [comment]
Thanks @cam72cam!

**Pull request merged**